genesyscloud/util_files_test.go

```diff
@@ -0,0 +1,125 @@
+package genesyscloud
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// TestS3UploadSuccess will simulate a successful uploading of a file to S3 using the S3 uploader.  It does not care about the actual YAML file contents and simply mocks out what gets retruned
+func TestS3UploadSuccess(t *testing.T) {
+	///Setting up variables
+	presignedURL := "/s3/presigned"
+	substitutions := make(map[string]interface{})
+	headers := make(map[string]string)
+	yamlFile := `inboundCall:
+					name: SimpleFinancialIvr
+					description: SimpleFinancialIvr
+					division: Home0349a372-0480-4879-b98e-8aace58b1bb7
+					startUpRef: "/inboundCall/menus/menu[Main Menu_10]"
+					defaultLanguage: en-us
+					supportedLanguages:
+						en-us:
+							defaultLanguageSkill:
+							noValue: true
+							textToSpeech:
+							defaultEngine:
+								voice: Jill`
+	fileReader := strings.NewReader(yamlFile)
+
+	// Create a mock HTTP server
+	mockServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
This is the money.  Go has a built in mock server where we can register handlers to mock out calls.  This allows us to write test cases where an API call is being made without having to do anything special to the actual underlying code (e.g. like exposing the httpClient).

Theoritically we could mock out the entire set of calls to our public API and never have a resource test actually touch Genesys Cloud.  This would significantly simplify our dependency on having a Genesys Cloud resource available.